File: scripts/algorithms/F/Fruit Into Baskets/Fruit Into Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int i=0, j=0, ans = 1;
        unordered_map<int, int>mp;
        while(j<fruits.size()){
            mp[fruits[j]]++;
            
            if(mp.size()<2){
                ans = max(ans, j-i+1);
                j++;
            }
            else if(mp.size()==2){
                ans = max(ans, j-i+1);
                j++;
            }
            else if(mp.size()>2){
                while(mp.size()>2){
                    mp[fruits[i]]--;
                    if(mp[fruits[i]]==0)
                        mp.erase(fruits[i]);
                    i++;
                }
                if(mp.size()==2){
                    ans = max(ans, j-i+1);
                }
                j++;
            }
        }
        return ans;
    }
};

File: scripts/algorithms/F/Fruit Into Baskets/Fruit Into Baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        ans=0
        fruitdict=defaultdict()
        stack=[]
        i,j=0,0

        while j<len(fruits):
            if fruits[j] not in fruitdict and len(fruitdict)<2:
                stack.append(fruits[j])
                fruitdict[fruits[j]]=j
                j+=1

            elif  fruits[j] in fruitdict:
                fruitdict[fruits[j]]=j
                j+=1

            else:  
                if fruitdict[stack[0]]>fruitdict[stack[1]]  :
                    i = fruitdict[stack[1]]+1
                    del fruitdict[stack[1]]
                    stack.pop()
                else:
                    i = fruitdict[stack[0]]+1
                    del fruitdict[stack[0]] 
                    stack.pop(0)              
            
            ans=max(ans,j-i)
        return ans